src/compose/service-manager.coffee

```diff
@@ -0,0 +1,276 @@
+
+	getAllByAppId: (appId) =>
+		@getAll("io.resin.app_id=#{appId}")
+		.filter (service) ->
+			service.appId == appId
+
+	stopAllByAppId: (appId) =>
+		Promise.map @getAllByAppId(appId), (service) =>
+			@kill(service, { removeContainer: false })
+
+	create: (service) =>
+		mockContainerId = @config.newUniqueKey()
+		@get(service)
+		.then ([ existingService ]) =>
+			return @docker.getContainer(existingService.containerId) if existingService?
+			conf = service.toContainerConfig()
+			@logger.logSystemEvent(logTypes.installService, { service })
+			@reportNewStatus(mockContainerId, service, 'Installing')
+			@docker.createContainer(conf)
+			.tap (container) =>
+				service.containerId = container.id
+				@logger.logSystemEvent(logTypes.installServiceSuccess, { service })
+		.catch (err) =>
+			@logger.logSystemEvent(logTypes.installServiceError, { service, error: err })
+			throw err
+		.finally =>
+			@reportChange(mockContainerId)
+
+	start: (service) =>
+		alreadyStarted = false
+		containerId = null
+		@create(service)
+		.tap (container) =>
+			containerId = container.id
+			@logger.logSystemEvent(logTypes.startService, { service })
+			@reportNewStatus(containerId, service, 'Starting')
+			container.start()
+			.catch (err) =>
+				statusCode = '' + err.statusCode
+				# 304 means the container was already started, precisely what we want :)
+				if statusCode is '304'
+					alreadyStarted = true
+					return
+
+				if statusCode is '500' and err.message?.trim?()?.match(/exec format error$/)
+					# Provide a friendlier error message for "exec format error"
+					@config.get('deviceType')
+					.then (deviceType) ->
+						throw new Error("Application architecture incompatible with #{deviceType}: exec format error")
+				else
+					# rethrow the same error
+					throw err
+			.catch (err) =>
+				# If starting the container failed, we remove it so that it doesn't litter
+				container.remove(v: true)
+				.finally =>
+					@logger.logSystemEvent(logTypes.startServiceError, { service, error: err })
+					throw err
+			.then =>
+				@logger.attach(@docker, container.id, service.serviceId)
+		.tap =>
+			if alreadyStarted
+				@logger.logSystemEvent(logTypes.startServiceNoop, { service })
+			else
+				@logger.logSystemEvent(logTypes.startServiceSuccess, { service })
+		.then (container) ->
+			service.running = true
+		.finally =>
+			@reportChange(containerId)
+
+	# Gets all existing containers that correspond to apps
+	getAll: (extraLabelFilters = []) =>
+		filters = label: [ 'io.resin.supervised' ].concat(extraLabelFilters)
+		@docker.listContainers({ all: true, filters })
+		.then (containers) =>
+			Promise.mapSeries containers, (container) =>
+				@docker.getContainer(container.Id).inspect()
+				.then(Service.fromContainer)
+
+	# Returns an array with the container(s) matching a service by appId, commit, image and environment
+	get: (service) =>
+		@getAll("io.resin.service_id=#{service.serviceId}")
+		.then (services = []) ->
```

Comment:
Why is this defaulting to an array? `getAll` should always return an array

Reply:
Fixed